ext/reflection/php_reflection.c

```diff
@@ -5715,6 +5715,106 @@ ZEND_METHOD(reflection_property, hasType)
 }
 /* }}} */
 
+/* {{{ proto public bool ReflectionProperty::hasDefaultValue()
+   Returns whether property has a default value */
+ZEND_METHOD(reflection_property, hasDefaultValue)
+{
+	reflection_object *intern;
+	property_reference *ref;
+	zend_property_info *prop_info;
+	zval *prop, prop_copy;
+	zend_class_entry *ce;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	GET_REFLECTION_OBJECT_PTR(ref);
+
+	prop_info = ref->prop;
+	ce = prop_info->ce;
+
+	prop = NULL;
+	if ((prop_info->flags & ZEND_ACC_STATIC) != 0) {
+		prop = &ce->default_static_members_table[prop_info->offset];
+		ZVAL_DEINDIRECT(prop);
+	} else if ((prop_info->flags & ZEND_ACC_STATIC) == 0) {
+		prop = &ce->default_properties_table[OBJ_PROP_TO_NUM(prop_info->offset)];
+	}
+
+	if (!prop || (ZEND_TYPE_IS_SET(prop_info->type) && Z_ISUNDEF_P(prop))) {
+		RETURN_FALSE;
+	}
+
+	/* copy: enforce read only access */
+	ZVAL_DEREF(prop);
+	ZVAL_COPY_OR_DUP(&prop_copy, prop);
+
+	/* this is necessary to make it able to work with default array
+	* properties, returned to user */
+	if (Z_TYPE(prop_copy) == IS_CONSTANT_AST) {
+		if (UNEXPECTED(zval_update_constant_ex(&prop_copy, ce) != SUCCESS)) {
+			RETURN_FALSE;
+		}
+	}
+
+	if (Z_TYPE_P(prop) != IS_NULL) {
```

Comment:
I think `hasDefaultValue()` should return true even if the default is null. We can't distinguish `public $foo;` and `public $foo = null;` here, and I'd say the latter at least clearly "has a default value".

Reply:
you can already use `$reflectionProperty->isDefault()` for this, this check is explicitly about default value that is not null, because we need to use this to test for statically (without instance) check for initialized property yes or no in context typed properties.

Reply:
`isDefault()` has no relation to this, it distinguishes between dynamic and declared properties.

Returning `false` for a null default value means that you have no way to distinguish between `public ?T $prop;` and `public ?T $prop = null;`.

Reply:
I changed this, it makes more sense and returns true now, for `public $foo;` and `public ?$foo = null;` but returns false for `public string $foo;` so it really returns false only if the variable is uninitialized and access (get before set) will throw an exception.